client/src/pages/spools/create.tsx

```diff
@@ -105,17 +113,83 @@ export const SpoolCreate: React.FC<IResourceComponentsProps & CreateOrCloneProps
             }
           />
         </Form.Item>
+        <Form.Item
+          hidden={true}
+          name={["used_weight"]}
+          initialValue={0}
+        >
+          <InputNumber
+            value={used_weight}
+          />
+        </Form.Item>
+
         <Form.Item
           label={t("spool.fields.used_weight")}
           help={t("spool.fields_help.used_weight")}
-          name={["used_weight"]}
-          rules={[
-            {
-              required: true,
-            },
-          ]}
+          // name={["used_weight"]}
+          initialValue={0}
+        >
+          <InputNumber
+            min={0}
+            addonAfter="g"
+            precision={1}
+            formatter={numberFormatter}
+            parser={numberParser}
+            value={used_weight}
+            onChange={(value) => {
+              set_used_weight(value ?? 0);
+              form.setFieldsValue(
+              {
+                used_weight: value ?? 0
+              });
+            }}
+          />
+        </Form.Item>
+		<Form.Item
+          label={t("spool.fields.remaining_weight")}
+          help={t("spool.fields_help.remaining_weight")}
+          // name={["remaining_weight"]}
+          initialValue={0}
         >
-          <InputNumber min={0} addonAfter="g" precision={1} formatter={numberFormatter} parser={numberParser} />
+          <InputNumber
+            min={0}
+            addonAfter="g"
+            precision={1}
+            formatter={numberFormatter}
+            parser={numberParser}
+            disabled={!(selected_filament?.weight || 0)}
+            value={(selected_filament?.weight || 0) ? (selected_filament?.weight || 0) - used_weight : 0}
+            onChange={(value) => {
+              set_used_weight((selected_filament?.weight || 0) - (value ?? 0));
+              form.setFieldsValue(
+                {
+                  used_weight: (selected_filament?.weight || 0) - (value ?? 0)
+                });
+            }}
+          />
+        </Form.Item>
+		<Form.Item
+          label={t("spool.fields.measured_weight")}
+          help={t("spool.fields_help.measured_weight")}
+          // name={["measured_weight"]}
+          initialValue={0}
+        >
+          <InputNumber
+            min={0}
+            addonAfter="g"
+            precision={1}
+            formatter={numberFormatter}
+            parser={numberParser}
+            disabled={!((selected_filament?.weight || 0) && (selected_filament?.spool_weight || 0))}
+            value={((selected_filament?.weight || 0) && (selected_filament?.spool_weight || 0)) ? (selected_filament?.weight || 0) - used_weight + (selected_filament?.spool_weight || 0) : 0}
```

Comment:
This is quite hard to read, can you perhaps split it up into a few variables? e.g. `const filament_weight = selected_filament?.weight || 0`

Reply:
ok i will, im a bit confused as to now veriables work in react, putting const in front of things that change feels wrong so i keep forgetting that you can here (as long as you dont have to change them in code later)

Reply:
This whole function re-runs everytime react detects that something has changed which warrants a re-render, such as the value of a state has changed, or your filament watch is triggered. The const variables are only constant for a single specific render.